### ISSUE TITLE
[attr-defined] for attribute defined in __new__

### ISSUE THREAD

**Issue by Hocnonsense** [python/mypy#15806]
**Bug Report**

This bug caused by defind an attribute in __new__ for a class. This bug had been reported in #982.

**To Reproduce**

```python
class MyObject:
    """
    A file that is either input or output of a rule.
    """

    __slots__ = ["a"]

    def __new__(cls, text):
        obj = object.__new__(cls)
        obj.a = text

        return obj


mo = MyObject("")
mo.a
```

**Expected Behavior**

```log
Success: no issues found in 1 source file
```

**Actual Behavior**

```log
2023/08/03.py:16: error: "MyObject" has no attribute "a"  [attr-defined]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: `mypy 1.4.1 (compiled: yes)`
- Mypy command-line flags: `mypy 2023/08/03.py`
- Mypy configuration options from `mypy.ini` (and other config files): None (default)
- Python version used: `Python 3.10.8`

**Comment by SanjayManjunath** [comment]
Hey @Hocnonsense, I found out that 'a' should be initialized in the '__init__' method since it's an instance attribute. Can you take a look at this and let me know if it's alright? https://mypy-play.net/?mypy=latest&python=3.10&gist=3e32e9112f3e5425c6aad2f8330e7ef8

**Comment by erictraut** [comment]
I think this is a duplicate of https://github.com/python/mypy/issues/1021.
